### PR TITLE
buildNpmPackage: support makeWrapperArgs

### PR DESCRIPTION
**Pull request by doronbehar**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: doc/languages-frameworks/javascript.section.md
Comment (lilyinstarlight) [inline doc/languages-frameworks/javascript.section.md:207]: Maybe make it clearer this is only used for `bin`s defined in package.json?

```suggestion
* `makeWrapperArgs`: Flags to pass to `makeWrapper` when wrapping package-defined `bin` entrypoints with `node`.
```

I'm open to wording suggestions, since it's a bit early this morning for me to write scrutable documentation, but this is just a suggestion of what I'm thinking

**Review by lilyinstarlight — COMMENTED** [review]
I've left one nit on the documentation, but this looks lovely otherwise, thank you!

**Comment by doronbehar** [comment]
Friendly ping @lilyinstarlight  (I want to add a package that will use this feature).

File: doc/languages-frameworks/javascript.section.md
Comment (lilyinstarlight) [inline doc/languages-frameworks/javascript.section.md:196]:
```suggestion
In the default `installPhase` set by `buildNpmPackage`, it uses `npm pack --json --dry-run` to decide what files to install in `$out/lib/node_modules/$name/`, where `$name` is the `name` string defined in the package's `package.json`. Additionally, the  `bin` and `man` keys in the source's `package.json` are used to decide what binaries and manpages are supposed to be installed. If these are not defined, `npm pack` may miss some files, and no binaries will be produced.
```

Maybe don't use `${src}` since if `sourceRoot` or `npmRoot` are set, the actual `package.json` location may be different

File: doc/languages-frameworks/javascript.section.md
Comment (lilyinstarlight) [inline doc/languages-frameworks/javascript.section.md:211]:
```suggestion
* `makeWrapperArgs`: Flags to pass to `makeWrapper`, added to executable calling the generated `.js` with `node` as an interpreter. These scripts are defined in `package.json`.
```

Typo?

File: doc/languages-frameworks/javascript.section.md
Comment (lilyinstarlight) [inline doc/languages-frameworks/javascript.section.md:232]:
```suggestion
It returns a derivation with all `package-lock.json` dependencies downloaded into `$out/`, usable as an npm cache.
```

Tiny voice nit (feel free to ignore that one, I don't feel that strongly) and clarifying it's an npm cache, not a generalized Node.js cache

**Review by lilyinstarlight — COMMENTED** [review]
I made some tiny nits below, but this looks lovely, thank you so much!

**Comment by doronbehar** [comment]
Pushed a rebase.

**Comment by doronbehar** [comment]
@lilyinstarlight I tested it now after your approval on the package I wanted to use this feature for, and I noticed that the quotes around `${makeWrapperArgs[@]}` in the shell script were not good. So that is the purpose of the last force push. Will merge afterwards when CI is green.

**Comment by lilyinstarlight** [comment]
> @lilyinstarlight I tested it now after your approval on the package I wanted to use this feature for, and I noticed that the quotes around `${makeWrapperArgs[@]}` in the shell script were not good. So that is the purpose of the last force push. Will merge afterwards when CI is green.

Oh then you probably need to magic the code to support both when it's a IFS-separated string and when it's a proper bash array. Maybe check how the other npm arg arrays work or check how the python wrapper handles that

**Comment by doronbehar** [comment]
Oh good idea, python's `wrap.sh` was useful. Pushed something now.

**Review by lilyinstarlight — APPROVED** [review]
Wrap stuff looks good to me! I left one document nit below, but it's pretty minor and doesn't need to be blocking

**Comment by doronbehar** [comment]
Haha it's OK. Thanks again for the approval. Pushed that suggestion. Will merge when CI is green.

**Pull request merged**
